STORM-1839: Storm spout implementation for Amazon Kinesis Streams.

priyank5485 opened the pull request.

harshach commented on external/storm-kinesis/src/main/java/org/apache/storm/kinesis/spout/Config.java:
```diff
@@ -0,0 +1,166 @@
+import java.io.Serializable;
+import java.util.Date;
+
+public class Config implements Serializable {
```
given that this is a public class going to be used by devs. Can we rename this as KinesisConfig

priyank5485 replied:
Will change it.

harshach commented on external/storm-kinesis/src/main/java/org/apache/storm/kinesis/spout/Config.java:
```diff
@@ -0,0 +1,166 @@
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
Do we need equals and hash code . Not sure where do we do comparisons of config object.

priyank5485 replied:
We dont need equals and hashcode. I will remove it.

harshach commented on external/storm-kinesis/src/main/java/org/apache/storm/kinesis/spout/ExponentialBackoffRetrier.java:
```diff
@@ -0,0 +1,164 @@
+        failCounts.put(messageId, ++failCount);
+        // if fail count is greater than maxRetries, discard or ack. for e.g. for maxRetries 3, 4 failures are allowed at maximum
+        if (failCount > maxRetries) {
+            LOG.debug("maxRetries reached so dropping " + messageId);
```
can you make this as warn as it can be useful to see this in the logs by default

priyank5485 replied:
I will change it

harshach commented on external/storm-kinesis/src/main/java/org/apache/storm/kinesis/spout/ExponentialBackoffRetrier.java:
```diff
@@ -0,0 +1,164 @@
+        // if reached so far, add it to the set of messages waiting to be retried with next retry time based on how many times it failed
+        retryTimes.put(messageId, getRetryTime(failCount));
+        retryMessageSet.add(messageId);
+        LOG.debug("Scheduled " + messageId + " for retry at " + retryTimes.get(messageId) + " and retry attempt " + failCount);
```
I recommend this to be at warn level as well.

priyank5485 replied:
I will change it

harshach commented on external/storm-kinesis/src/main/java/org/apache/storm/kinesis/spout/ExponentialBackoffRetrier.java:
```diff
@@ -0,0 +1,164 @@
+        LOG.debug("Handling failed message " + messageId);
+        // if maxRetries is 0, dont retry and return false as per interface contract
+        if (maxRetries == 0) {
+            LOG.debug("maxRetries set to 0. Hence not queueing " + messageId);
```
warn here as well

priyank5485 replied:
I will change it.

harshach commented on external/storm-kinesis/src/main/java/org/apache/storm/kinesis/spout/ExponentialBackoffRetrier.java:
```diff
@@ -0,0 +1,164 @@
+public class ExponentialBackoffRetrier implements FailedMessageRetryHandler, Serializable {
+    private static final Logger LOG = LoggerFactory.getLogger(ExponentialBackoffRetrier.class);
+    // Wait interfal for retrying after first failure
+    private final Long initialDelayMillis;
```
should we have any defaults for these variables.

priyank5485 replied:
ExponentialBackoffRetrier has two constructors. The no args constructor will set the default values for these variables.

harshach commented on external/storm-kinesis/src/main/java/org/apache/storm/kinesis/spout/ExponentialBackoffRetrier.java:
```diff
@@ -0,0 +1,164 @@
+    // map to track next retry time for each kinesis message that failed
+    private Map<KinesisMessageId, Long> retryTimes = new HashMap<>();
+    // sorted set of records to be retrued based on retry time. earliest retryTime record comes first
+    private SortedSet<KinesisMessageId> retryMessageSet = new TreeSet<>(new RetryTimeComparator());
```
are we going to store the same message in retryMessageSet and retryTimes?

priyank5485 replied:
No. Messages are not stored anywhere in ExponentialBackoffRetrier. retryTimes is just used to implement the comparator interface that will be used by retryMessageSet TreeSet to order the messages on their next retry time. You can check out the RetryTimeComparator in this class

harshach commented on external/storm-kinesis/src/main/java/org/apache/storm/kinesis/spout/KinesisConnectionInfo.java:
```diff
@@ -0,0 +1,137 @@
+        this.recordsLimit = recordsLimit;
+        this.region = region;
+
+        this.credentialsProvider = null;
```
we don't need to explicitly assign a null to these.

priyank5485 replied:
Will change it

harshach commented on external/storm-kinesis/src/main/java/org/apache/storm/kinesis/spout/KinesisConnectionInfo.java:
```diff
@@ -0,0 +1,137 @@
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
we probably can remove these equals, hash here as well.

priyank5485 replied:
Will change it

harshach commented on external/storm-kinesis/src/main/java/org/apache/storm/kinesis/spout/KinesisRecordsManager.java:
```diff
@@ -0,0 +1,566 @@
+        LOG.debug("No failed record to emit for now. Hence will try to emit new records");
+        // if maximum uncommitted records count has reached, so dont emit any new records and return
+        if (!(getUncommittedRecordsCount() < config.getMaxUncommittedRecords())) {
+            LOG.debug("maximum uncommitted records count has reached. so not emitting any new records and returning");
```
this should be warn

harshach commented on external/storm-kinesis/src/main/java/org/apache/storm/kinesis/spout/KinesisRecordsManager.java:
```diff
@@ -0,0 +1,566 @@
+        }
+        // early return as no shard is assigned - probably because number of executors > number of shards
+        if (toEmitPerShard.isEmpty()) {
+            LOG.debug("No shard is assigned to this task. Hence not emitting any tuple.");
```
same here

harshach commented on external/storm-kinesis/src/main/java/org/apache/storm/kinesis/spout/KinesisRecordsManager.java:
```diff
@@ -0,0 +1,566 @@
+                if (emitFailedRecord(collector, failedMessageId)) {
+                    failedPerShard.get(failedMessageId.getShardId()).remove(failedSequenceNumber);
+                    config.getFailedMessageRetryHandler().failedMessageEmitted(failedMessageId);
+                    return;
```
if we can make multiple returns into one or two we should try.

priyank5485 replied:
I thought about it but if i try to reduce the return statements it will create more if else statements. If its not a big issue we can keep as is.

harshach commented on external/storm-kinesis/src/main/java/org/apache/storm/kinesis/spout/KinesisRecordsManager.java:
```diff
@@ -0,0 +1,566 @@
+        emitNewRecord(collector);
+    }
+
+    void ack (KinesisMessageId kinesisMessageId) {
```
if we can document via comments on this ack logic that would be great.

priyank5485 replied:
I documented the logic in the commit method since thats where it writes to zookeeper. Let me know if that is not enough or if i should move it to somewhere else.

harshach replied:
@priyank5485 looks like this method is going through a switch statement of sort, i.e its checking bunch if conditions and doing actions on that part. It will be great if you can add why you are doing it. For the future reference it will be important as this part is most important on when to ack a message.

harshach commented on external/storm-kinesis/README.md:
```diff
@@ -0,0 +1,139 @@
+3 types are supported - TRIM_HORIZON(beginning of shard), LATEST and AT_TIMESTAMP. By default this argument is ignored if state for shards 
+is found in zookeeper. Hence they will apply the first time a topology is started. If you want to use any of these in subsequent runs of the topology, you 
+will need to clear the state of zookeeper node used for storing sequence numbers
+#### `RecordToTupleMapper` recordToTupleMapper
```
line break

priyank5485 replied:
Will change it

harshach commented on external/storm-kinesis/src/main/java/org/apache/storm/kinesis/spout/KinesisRecordsManager.java:
```diff
@@ -0,0 +1,566 @@
+                shardIteratorPerFailedMessage.put(kinesisMessageId, getRecordsResult.getNextShardIterator());
+                if (records.size() == 0) {
+                    LOG.debug("No records returned from kinesis. Hence sleeping for 1 second");
+                    Thread.sleep(1000);
```
should this be configurable ? or probably a candidate for exponential back off . we can address this in follow-up jira

priyank5485 replied:
I would prefer this to be not configurable. Reason is kinesis api somewhere mentioned in their documentation that if you dont receive any records it is good to sleep for a second. Reason is that they some limits for GetRecords request per second and it will throw an exception if that limit is reached or violated. Although I am catching that exception I wanted to avoid that. Making it configurable and go beyond 1 second can slow down the spout in some cases as well. For example, sometimes it takes multiple requests to get to the portion of the shard where the message at a sequence number you want to read is present. In that case it does not make sense to backoff in an exponential fashion. You just want to make sure you get there sooner without violating the limit.

harshach commented on external/storm-kinesis/src/main/java/org/apache/storm/kinesis/spout/KinesisRecordsManager.java:
```diff
@@ -0,0 +1,566 @@
+                }
+            }
+        } catch (InterruptedException ie) {
+            LOG.debug("Thread interrupted while sleeping", ie);
```
lets make these warns for all the exceptions unless you think this might result lot of noise. Even than it will be helpful to print them given that these are critical errors.

harshach commented on external/storm-kinesis/src/main/java/org/apache/storm/kinesis/spout/KinesisRecordsManager.java:
```diff
@@ -0,0 +1,566 @@
+        } catch (ProvisionedThroughputExceededException pe) {
+            try {
+                LOG.debug("ProvisionedThroughputExceededException occured. Check your limits. Sleeping for 1 second.", pe);
+                Thread.sleep(1000);
```
same as before

harshach commented on external/storm-kinesis/src/main/java/org/apache/storm/kinesis/spout/KinesisRecordsManager.java:
```diff
@@ -0,0 +1,566 @@
+        kinesisClient.shutdown();
+    }
+
+}
```
This is big file thats doing lot of things. We should some of these methods into Utils class or something  like zk related methods into its own class.

harshach replied:
@priyank5485 do you think its possible to move some of these methods into its Utils file or some sorts.

harshach commented on external/storm-kinesis/src/main/java/org/apache/storm/kinesis/spout/ZkInfo.java:
```diff
@@ -0,0 +1,153 @@
+     * Default constructor that uses defaults for a local setup
+     */
+    public ZkInfo () {
+        this("localhost:2181", "/kinesisOffsets", 20000, 15000, 10000L, 3, 2000);
```
I think this is useful for testing but I can see users tripping over this :) so can we remove this.

priyank5485 replied:
Will remove it

harshach commented on external/storm-kinesis/src/main/java/org/apache/storm/kinesis/spout/ZkInfo.java:
```diff
@@ -0,0 +1,153 @@
+                '}';
+    }
+
+    @Override
```
remove these as well.

harshach commented:
@priyank5485 looks great!. Mostly nit-picks. I'll go over the retry handle one more time and add any comments if needed.

priyank5485 commented:
@harshach Thanks for the review. I have addressed your comments and pushed a commit.

harshach commented:
@priyank5485 left couple of comments

priyank5485 commented:
@harshach I added some more comments in ack method and commit method for acking logic. I also moved code interacting with zk and kinesis in to its own class to make KinesisRecordsManager smaller. Please review when you get a chance

harshach commented on external/storm-kinesis/src/main/java/org/apache/storm/kinesis/spout/KinesisRecordsManager.java:
```diff
@@ -0,0 +1,449 @@
+                shardIteratorPerFailedMessage.put(kinesisMessageId, getRecordsResult.getNextShardIterator());
+                if (records.size() == 0) {
+                    LOG.warn("No records returned from kinesis. Hence sleeping for 1 second");
+                    Thread.sleep(1000);
```
Sorry missed these earlier. Any reason for it to sleep exactly one sec or should it be configurable.

priyank5485 replied:
I would prefer this to be not configurable. Reason is kinesis api somewhere mentioned in their documentation that if you dont receive any records it is good to sleep for a second. Reason is that they some limits for GetRecords request per second and it will throw an exception if that limit is reached or violated. Although I am catching that exception I wanted to avoid that. Making it configurable and go beyond 1 second can slow down the spout in some cases as well. For example, sometimes it takes multiple requests to get to the portion of the shard where the message at a sequence number you want to read is present. In that case it does not make sense to backoff in an exponential fashion. You just want to make sure you get there sooner without violating the limit.

harshach commented:
+1. Build failures not related to the patch.

revans2 commented:
@harshach actually some of the build errors are related see https://issues.apache.org/jira/browse/STORM-2021
